hydra/utils.py

```diff
@@ -12,38 +13,82 @@
 log = logging.getLogger(__name__)
 
 
+def _safeimport(path: str) -> Optional[ModuleType]:
+    """
+    Import a module; handle errors; return None if the module isn't found.
+    This is a typed simplified version of the `pydoc` function `safeimport`.
+    """
+    from importlib import import_module
+
+    try:
+        module = import_module(path)
+    except ImportError as e:
+        if e.name == path:
+            return None
+        else:
+            log.error(f"Error importing module: {path}")
+            raise e
+    except Exception as e:
+        log.error(f"Non-ImportError while importing module {path}: {e}")
+        raise ValueError(f"Non-ImportError while importing module {path}: {e}")
+    for part in path.replace(module.__name__, "").split("."):
+        if not hasattr(module, part):
+            break
+        module = getattr(module, part)
+    return module
+
+
+def _locate(path: str) -> ModuleType:
+    """
+    Locate an object by name or dotted path, importing as necessary.
+    This is similar to the pydoc function `locate`, except that it checks for
+    the module from the given path from back to front.
+    """
+    parts = [part for part in path.split(".") if part]
+    module = None
+    for n in reversed(range(len(parts))):
+        try:
+            module = _safeimport(".".join(parts[:n]))
+        except Exception:
+            continue
+        if module:
+            break
+    if module:
+        obj = module
+    else:
+        log.error(f"Module not found: {path}")
+        raise ValueError(f"Module not found: {path}")
+    for part in parts[n:]:
+        if not hasattr(obj, part):
+            log.error(
+                f"Error finding attribute ({part}) in class ({obj.__name__}): {path}"
+            )
+            raise ValueError(
+                f"Error finding attribute ({part}) in class ({obj.__name__}): {path}"
+            )
+        obj = getattr(obj, part)
+    return obj
+
+
 def get_method(path: str) -> type:
     return get_class(path)
 
 
 def get_class(path: str) -> type:
     try:
-        from importlib import import_module
-
-        module_path, _, class_name = path.rpartition(".")
-        mod = import_module(module_path)
-        try:
-            klass: type = getattr(mod, class_name)
-        except AttributeError:
-            raise ImportError(
-                "Class {} is not in module {}".format(class_name, module_path)
-            )
+        klass = cast(type, _locate(path))
```

Comment:
try : 

```python
t = _locate(path)
assert isinstance(t, type)
return t
```

Reply:
I tried this but it didn't work.  The return types of static methods and class methods were different than class objects.  I tried explicitly typing `t` with `t: type`, but assert statement still failed.  I also tried changing the assert statement to `Callable` instead of `type`, but then I recieved a failure from the mypy check.  In this case, all the tests passed.

I also tried changing all of the return types to `ModuleType` and adjust the function signature of `_instantiate_class` accordingly, but this caused more mypy errors and some of the tests started failing as well.

Reply:
_locate signature suggests that the return type is ModuleType.
Since this function is now used for more than returning classes, maybe we need to refactor things such that it only get called for actual classes (by the user) and the internal implementation operates on ModuleType.

Does it make sense?
btw, can you join the chat?

Reply:
I am in Europe so I can join a chat today (for me, probably tomorrow for you) if you're free.  I just registered for the chat, so I'll be in there.